isic/core/services/collection/doi.py

```diff
@@ -91,20 +102,29 @@ def _datacite_create_doi(doi: dict) -> None:
 def collection_create_doi(*, user: User, collection: Collection) -> Doi:
     collection_check_create_doi_allowed(user=user, collection=collection)
     doi_id = collection_generate_random_doi_id()
+    draft_doi = collection_build_draft_doi(doi_id=doi_id)
     doi = collection_build_doi(collection=collection, doi_id=doi_id)
 
-    try:
-        _datacite_create_doi(doi)
-    except HTTPError as e:
-        logger.error(e)
-        raise ValidationError("Something went wrong creating the DOI.")
-    else:
-        with transaction.atomic():
-            doi = Doi(id=doi_id, creator=user, url=f"https://doi.org/{doi_id}")
-            doi.full_clean()
-            doi.save()
-            collection_lock(collection=collection)
-            collection_update(collection=collection, doi=doi, ignore_lock=True)
-        logger.info("User %d created DOI %s for collection %d", user.id, doi.id, collection.id)
-
+    with transaction.atomic():
+        # First, create the local DOI record to validate uniqueness within our known set
+        doi = Doi(id=doi_id, creator=user, url=f"https://doi.org/{doi_id}")
+        doi.full_clean()
+        doi.save()
+
+        # Lock the collection, set the DOI on it
+        collection_lock(collection=collection)
+        collection_update(collection=collection, doi=doi, ignore_lock=True)
+
+        try:
+            # Reserve the DOI using the draft mechanism
+            _datacite_create_doi(draft_doi)
+        except HTTPError as e:
+            logger.error(e)
+            # Transaction will rollback, nothing in our database will change
+            raise ValidationError("Something went wrong creating the DOI.")
+        else:
+            logger.info("User %d created DOI %s for collection %d", user.id, doi.id, collection.id)
+
+    # Convert to a published DOI. Only question now is what happens if this fails.
+    _datacite_create_doi(doi)
```

Comment:
PTAL at this revised order of operations. If we wanted to close the loop here to make this truly robust, I would suggest adding either a new status column onto the `Doi` table to be able to lookup which ones are stuck in the draft state, or else a separate relation `DraftDoi` to store them and retry the publish operation later.

The other failure mode to consider is if the draft DOI is created, but the transaction commit fails. You'd be left with an orphaned draft DOI, but maybe that's such a remote possibility and of no real consequence that we don't care. Our DOI space is < 1M elements, so we obviously don't want to be too flippantly orphaning space within that.

Reply:
> The other failure mode to consider is if the draft DOI is created, but the transaction commit fails. You'd be left with an orphaned draft DOI, but maybe that's such a remote possibility and of no real consequence that we don't care.

I don't think it matters. The big concern I had was creating duplicate non-draft DOIs, since those can't be deleted and instead have to be manually tracked down and delisted. This fixes that.

Reply:
Do you want me to create the database record for draft status, or is this an infrequent enough operation that you wouldn't mind fixing any issues manually?

Reply:
Manually is fine.